File: backend/api/customers.php
<?php
require_once __DIR__ . '/../config/bootstrap.php';
require_once __DIR__ . '/../utils/pool.php';
require_once __DIR__ . '/../utils/mailer.php';

$method = $_SERVER['REQUEST_METHOD'];
$db = getDB();

function createDefaultSegments($db, $customerId) {
    $defaults = [
        ['Hauptgewinn', '#FFD700', 1, 'gold'],
        ['Gewinn', '#4CAF50', 3, 'nature'],
        ['Trostpreis', '#2196F3', 5, 'water'],
        ['Nochmal drehen', '#FF9800', 3, 'sand'],
    ];
    $stmt = $db->prepare("INSERT INTO segments (customer_id, name, color, win_text, weight, sort_order, max_count, theme) VALUES (?, ?, ?, ?, ?, ?, ?, ?)");
    foreach ($defaults as $i => $seg) {
        $stmt->execute([$customerId, $seg[0], $seg[1], $seg[0], 100, $i, $seg[2], $seg[3]]);
    }
}

function createDefaultSettings($db, $customerId) {
    $defaults = [
        'primary_color' => '#1E3A8A',
        'secondary_color' => '#F6A7C4',
        'accent_color' => '#D4AF37',
        'font_family' => 'Montserrat',
        'wheel_title' => 'Glücksrad',
        'campaign_status' => 'running',
        'theme' => 'dove',
        'total_spins_limit' => '1000'
    ];
    $stmt = $db->prepare("INSERT OR IGNORE INTO settings (customer_id, setting_key, setting_value) VALUES (?, ?, ?)");
    foreach ($defaults as $key => $value) {
        $stmt->execute([$customerId, $key, $value]);
    }
}

// GET /api/customers - Liste aller Kunden
if ($method === 'GET') {
    requireSuperAdmin();
    $stmt = $db->query("
        SELECT 
            c.id, c.company_name, c.contact_name, c.email, c.logo, c.is_active, c.subdomain, c.created_at,
            (SELECT COUNT(*) FROM users u WHERE u.customer_id = c.id) AS user_count,
            (SELECT COUNT(*) FROM spins s WHERE s.customer_id = c.id) AS spin_count
        FROM customers c
        ORDER BY c.company_name
    ");
    jsonResponse($stmt->fetchAll());
}

// POST /api/customers - Neuen Kunden anlegen
if ($method === 'POST') {
    requireSuperAdmin();
    validateCsrf();

    $data = json_decode(file_get_contents('php://input'), true);
    $companyName = sanitizePlainText($data['company_name'] ?? '');
    $contactName = sanitizePlainText($data['contact_name'] ?? '');
    $email = sanitizePlainText($data['email'] ?? '');
    $subdomain = sanitizePlainText($data['subdomain'] ?? '');

    if (empty($companyName)) {
        jsonResponse(['error' => 'Firmenname ist erforderlich'], 400);
    }
    if (empty($email) || !filter_var($email, FILTER_VALIDATE_EMAIL)) {
        jsonResponse(['error' => 'Gültige E-Mail ist erforderlich'], 400);
    }

    // Explizite Prüfung, damit wir eine präzise Fehlermeldung zurückgeben können
    $stmt = $db->prepare("SELECT id FROM users WHERE email = ?");
    $stmt->execute([$email]);
    if ($stmt->fetch()) {
        jsonResponse(['error' => 'E-Mail wird bereits verwendet'], 409);
    }

    if (!empty($subdomain)) {
        $stmt = $db->prepare("SELECT id FROM customers WHERE subdomain = ?");
        $stmt->execute([$subdomain]);
        if ($stmt->fetch()) {
            jsonResponse(['error' => 'Subdomain wird bereits verwendet'], 409);
        }
    }

    // Zufälliges 8-Zeichen Passwort für den automatischen Kunden-Admin generieren
    $plainPassword = substr(str_shuffle('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*'), 0, 8);
    $passwordHash = password_hash($plainPassword, PASSWORD_BCRYPT);

    $db->beginTransaction();
    try {
        $stmt = $db->prepare("INSERT INTO customers (company_name, contact_name, email, subdomain, is_active) VALUES (?, ?, ?, ?, 1)");
        $stmt->execute([$companyName, $contactName, $email, $subdomain]);
        $customerId = $db->lastInsertId();

        createDefaultSegments($db, $customerId);
        createDefaultSettings($db, $customerId);
        generateSpinPool($db, $customerId);

        // Automatisch einen Admin-Benutzer für den Kunden anlegen
        $stmt = $db->prepare("INSERT INTO users (email, username, password_hash, role, customer_id, is_active) VALUES (?, ?, ?, 'customer_admin', ?, 1)");
        $stmt->execute([$email, $email, $passwordHash, $customerId]);

        $db->commit();

        // Zugangsdaten per E-Mail an den Kunden senden
        $mailResult = sendCustomerCredentials($email, $contactName ?: $companyName, $email, $plainPassword);

        jsonResponse([
            'success' => true,
            'id' => $customerId,
            'message' => 'Kunde angelegt',
            'email_sent' => $mailResult['success'],
            'email_error' => $mailResult['success'] ? null : $mailResult['error']
        ]);
    } catch (PDOException $e) {
        $db->rollBack();
        if (strpos($e->getMessage(), 'UNIQUE') !== false) {
            jsonResponse(['error' => 'E-Mail oder Subdomain wird bereits verwendet', 'details' => $e->getMessage()], 409);
        }
        jsonResponse(['error' => 'Fehler beim Anlegen: ' . $e->getMessage()], 500);
    } catch (Exception $e) {
        $db->rollBack();
        jsonResponse(['error' => 'Fehler beim Anlegen: ' . $e->getMessage()], 500);
    }
}

// PUT /api/customers?id=... - Kunde bearbeiten
if ($method === 'PUT') {
    requireSuperAdmin();
    validateCsrf();

    $id = intval($_GET['id'] ?? 0);
    if (!$id) jsonResponse(['error' => 'ID erforderlich'], 400);

    $data = json_decode(file_get_contents('php://input'), true);
    $companyName = sanitizePlainText($data['company_name'] ?? '');
    $contactName = sanitizePlainText($data['contact_name'] ?? '');
    $email = sanitizePlainText($data['email'] ?? '');
    $subdomain = sanitizePlainText($data['subdomain'] ?? '');
    $isActive = isset($data['is_active']) ? intval($data['is_active']) : null;

    $updates = [];
    $params = [];
    if ($companyName !== '') { $updates[] = 'company_name = ?'; $params[] = $companyName; }
    if (isset($data['contact_name'])) { $updates[] = 'contact_name = ?'; $params[] = $contactName; }
    if (isset($data['email'])) { $updates[] = 'email = ?'; $params[] = $email; }
    if (isset($data['subdomain'])) { $updates[] = 'subdomain = ?'; $params[] = $subdomain; }
    if ($isActive !== null) { $updates[] = 'is_active = ?'; $params[] = $isActive; }

    if (empty($updates)) {
        jsonResponse(['error' => 'Keine Felder zum Aktualisieren'], 400);
    }

    $params[] = $id;
    $sql = "UPDATE customers SET " . implode(', ', $updates) . " WHERE id = ?";
    $stmt = $db->prepare($sql);
    $stmt->execute($params);

    // E-Mail des automatischen Kunden-Admin-Benutzers synchronisieren
    if (isset($data['email'])) {
        $stmt = $db->prepare("UPDATE users SET email = ?, username = ? WHERE customer_id = ? AND role = 'customer_admin'");
        $stmt->execute([$email, $email, $id]);
    }

    jsonResponse(['success' => true, 'message' => 'Kunde aktualisiert']);
}

// DELETE /api/customers?id=... - Kunde löschen
if ($method === 'DELETE') {
    requireSuperAdmin();
    validateCsrf();

    $id = intval($_GET['id'] ?? 0);
    if (!$id) jsonResponse(['error' => 'ID erforderlich'], 400);

    $db->beginTransaction();
    try {
        // Abhängige Daten in korrekter Reihenfolge löschen (Foreign Keys)
        $tables = ['spin_pool', 'spins', 'segments', 'settings', 'campaign_archives', 'users'];
        foreach ($tables as $table) {
            $stmt = $db->prepare("DELETE FROM {$table} WHERE customer_id = ?");
            $stmt->execute([$id]);
        }

        $stmt = $db->prepare("DELETE FROM customers WHERE id = ?");
        $stmt->execute([$id]);

        if ($stmt->rowCount() === 0) {
            $db->rollBack();
            jsonResponse(['error' => 'Kunde nicht gefunden'], 404);
        }

        $db->commit();
        jsonResponse(['success' => true, 'message' => 'Kunde gelöscht']);
    } catch (Exception $e) {
        $db->rollBack();
        jsonResponse(['error' => 'Fehler beim Löschen: ' . $e->getMessage()], 500);
    }
}

File: server/internal/httpapi/customers.go
package httpapi

import (
	"crypto/rand"
	"encoding/json"
	"errors"
	"math/big"
	"net/http"
	"strconv"
	"strings"

	"github.com/go-chi/chi/v5"
	"gluecksrad/server/internal/mail"
	"gluecksrad/server/internal/models"
	"gluecksrad/server/internal/pool"
	"gorm.io/gorm"
	"gorm.io/gorm/clause"
)

// routeCustomers montiert die Customers-Endpunkte.
//
// Auth-Regeln (aus customers.php):
//   - Alle Methoden → requireSuperAdmin (nur Super-Admin verwaltet Kunden).
//
// Routen:
//   - GET    /api/customers        → Liste aller Kunden
//   - POST   /api/customers        → Neuen Kunden anlegen (inkl. Default-Kampagne + User)
//   - PUT    /api/customers?id=    → Kunde bearbeiten (auch export_enabled)
//   - DELETE /api/customers?id=    → Kunde löschen (cascade)
func (s *Server) routeCustomers(r chi.Router) {
	r.Group(func(pr chi.Router) {
		pr.Use(s.requireAuth)
		pr.Use(s.requireSuperAdmin)
		pr.Get("/api/customers", s.handleGetCustomers)
		pr.Post("/api/customers", s.handlePostCustomer)
		pr.Put("/api/customers", s.handlePutCustomer)
		pr.Delete("/api/customers", s.handleDeleteCustomer)
	})
}

// customerListRow ist die flache Projektion die GET /api/customers zurückgibt.
type customerListRow struct {
	ID            uint64 `json:"id"`
	CompanyName   string `json:"company_name"`
	ContactName   string `json:"contact_name"`
	Email         string `json:"email"`
	Logo          string `json:"logo"`
	IsActive      bool   `json:"is_active"`
	Subdomain     string `json:"subdomain"`
	ExportEnabled bool   `json:"export_enabled"`
	CampaignLimit int    `json:"campaign_limit"`
	CreatedAt     string `json:"created_at"`
	UserCount     int    `json:"user_count"`
	CampaignCount int    `json:"campaign_count"`
}

// GET /api/customers
// Liste aller Kunden mit Anzahl User und Kampagnen (via Subquery).
func (s *Server) handleGetCustomers(w http.ResponseWriter, r *http.Request) {
	var rows []customerListRow
	err := s.DB.Raw(`
		SELECT
			c.id,
			c.company_name,
			c.contact_name,
			c.email,
			c.logo,
			c.is_active,
			c.subdomain,
			c.export_enabled,
			c.campaign_limit,
			c.created_at,
			(SELECT COUNT(*) FROM users u WHERE u.customer_id = c.id)     AS user_count,
			(SELECT COUNT(*) FROM campaigns ca WHERE ca.customer_id = c.id) AS campaign_count
		FROM customers c
		ORDER BY c.company_name
	`).Scan(&rows).Error
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Datenbankfehler: "+err.Error())
		return
	}
	if rows == nil {
		rows = []customerListRow{}
	}
	writeJSON(w, http.StatusOK, rows)
}

// postCustomerRequest ist der erwartete JSON-Body für POST /api/customers.
type postCustomerRequest struct {
	CompanyName string `json:"company_name"`
	ContactName string `json:"contact_name"`
	Email       string `json:"email"`
	Subdomain   string `json:"subdomain"`
}

// POST /api/customers
// Legt einen neuen Kunden an:
//  1. Validierung (email, company_name, Duplikat-Check)
//  2. Transaktion: customer → default-Kampagne → default-Segmente → default-Settings
//     → customer_admin-User → pool.GenerateForCampaign
//  3. Nach Commit: mail.SendCredentials; bei Fehler → password im Response-Body.
func (s *Server) handlePostCustomer(w http.ResponseWriter, r *http.Request) {
	var req postCustomerRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "Ungültiger JSON-Body")
		return
	}

	req.CompanyName = strings.TrimSpace(req.CompanyName)
	req.ContactName = strings.TrimSpace(req.ContactName)
	req.Email = strings.TrimSpace(req.Email)
	req.Subdomain = strings.TrimSpace(req.Subdomain)

	if req.CompanyName == "" {
		writeError(w, http.StatusBadRequest, "Firmenname ist erforderlich")
		return
	}
	if req.Email == "" || !isValidEmail(req.Email) {
		writeError(w, http.StatusBadRequest, "Gültige E-Mail ist erforderlich")
		return
	}

	// Duplikat-Check E-Mail
	var existingUser models.User
	if err := s.DB.Where("email = ?", req.Email).First(&existingUser).Error; err == nil {
		writeError(w, http.StatusConflict, "E-Mail wird bereits verwendet")
		return
	}

	// Duplikat-Check Subdomain
	if req.Subdomain != "" {
		var existingCustomer models.Customer
		if err := s.DB.Where("subdomain = ?", req.Subdomain).First(&existingCustomer).Error; err == nil {
			writeError(w, http.StatusConflict, "Subdomain wird bereits verwendet")
			return
		}
	}

	// Zufälliges 8-Zeichen Passwort generieren (krypto-sicher)
	plainPassword, err := generatePassword(8)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Fehler beim Generieren des Passworts")
		return
	}

	passwordHash, err := hashPassword(plainPassword)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Fehler beim Hashen des Passworts")
		return
	}

	var customerID uint64

	txErr := s.DB.Transaction(func(tx *gorm.DB) error {
		// 1. Kunden anlegen
		customer := models.Customer{
			CompanyName: req.CompanyName,
			ContactName: req.ContactName,
			Email:       req.Email,
			Subdomain:   req.Subdomain,
			IsActive:    true,
		}
		if err := tx.Create(&customer).Error; err != nil {
			return err
		}
		customerID = customer.ID

		// 2. Standard-Kampagne anlegen (status: draft)
		displayName := req.CompanyName + " – Kampagne 1"
		campaign := models.Campaign{
			CustomerID:     customerID,
			Name:           displayName,
			Status:         "draft",
			OnEmpty:        "end",
			EstimatedSpins: 100,
		}
		if err := tx.Create(&campaign).Error; err != nil {
			return err
		}
		campaignID := campaign.ID

		// 3. Standard-Segmente (scoped per campaign_id wie im Go-Modell)
		if err := createDefaultSegments(tx, campaignID); err != nil {
			return err
		}

		// 4. Standard-Settings (scoped per campaign_id)
		if err := createDefaultSettings(tx, campaignID); err != nil {
			return err
		}

		// 5. Spin-Pool generieren
		if _, err := pool.GenerateForCampaign(tx, campaignID); err != nil {
			return err
		}

		// 6. Kunden-Admin-User anlegen
		user := models.User{
			Email:        req.Email,
			Username:     req.Email,
			PasswordHash: passwordHash,
			Role:         "customer_admin",
			CustomerID:   &customerID,
			IsActive:     true,
		}
		if err := tx.Create(&user).Error; err != nil {
			return err
		}

		return nil
	})

	if txErr != nil {
		if strings.Contains(txErr.Error(), "unique") || strings.Contains(txErr.Error(), "UNIQUE") ||
			strings.Contains(txErr.Error(), "duplicate") {
			writeError(w, http.StatusConflict, "E-Mail oder Subdomain wird bereits verwendet")
			return
		}
		writeError(w, http.StatusInternalServerError, "Fehler beim Anlegen: "+txErr.Error())
		return
	}

	// Nach Commit: Zugangsdaten per E-Mail senden
	toName := req.ContactName
	if toName == "" {
		toName = req.CompanyName
	}
	mailErr := mail.SendCredentials(req.Email, toName, req.Email, plainPassword)
	emailSent := mailErr == nil

	resp := map[string]any{
		"success":    true,
		"id":         customerID,
		"email_sent": emailSent,
	}
	if !emailSent {
		// Fallback: Passwort im Klartext zurückgeben, damit Super-Admin es manuell weitergibt
		resp["password"] = plainPassword
		resp["email_error"] = mailErr.Error()
	}

	writeJSON(w, http.StatusCreated, resp)
}

// putCustomerRequest ist der JSON-Body für PUT /api/customers?id=.
// Pointer-Felder erlauben Partial-Update (nur gesetzte Felder werden übernommen).
type putCustomerRequest struct {
	CompanyName   *string `json:"company_name"`
	ContactName   *string `json:"contact_name"`
	Email         *string `json:"email"`
	Subdomain     *string `json:"subdomain"`
	IsActive      *bool   `json:"is_active"`
	ExportEnabled *bool   `json:"export_enabled"`
	CampaignLimit *int    `json:"campaign_limit"`
}

// PUT /api/customers?id=
// Partial-Update; synchronisiert auch die E-Mail des customer_admin-Users.
func (s *Server) handlePutCustomer(w http.ResponseWriter, r *http.Request) {
	idStr := r.URL.Query().Get("id")
	if idStr == "" {
		writeError(w, http.StatusBadRequest, "ID erforderlich")
		return
	}
	id, err := strconv.ParseUint(idStr, 10, 64)
	if err != nil || id == 0 {
		writeError(w, http.StatusBadRequest, "Ungültige ID")
		return
	}

	var req putCustomerRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "Ungültiger JSON-Body")
		return
	}

	updates := map[string]any{}
	if req.CompanyName != nil && *req.CompanyName != "" {
		updates["company_name"] = *req.CompanyName
	}
	if req.ContactName != nil {
		updates["contact_name"] = *req.ContactName
	}
	if req.Email != nil {
		if *req.Email != "" && !isValidEmail(*req.Email) {
			writeError(w, http.StatusBadRequest, "Ungültige E-Mail-Adresse")
			return
		}
		updates["email"] = *req.Email
	}
	if req.Subdomain != nil {
		updates["subdomain"] = *req.Subdomain
	}
	if req.IsActive != nil {
		updates["is_active"] = *req.IsActive
	}
	if req.ExportEnabled != nil {
		updates["export_enabled"] = *req.ExportEnabled
	}
	if req.CampaignLimit != nil {
		updates["campaign_limit"] = *req.CampaignLimit
	}

	if len(updates) == 0 {
		writeError(w, http.StatusBadRequest, "Keine Felder zum Aktualisieren")
		return
	}

	txErr := s.DB.Transaction(func(tx *gorm.DB) error {
		result := tx.Model(&models.Customer{}).Where("id = ?", id).Updates(updates)
		if result.Error != nil {
			return result.Error
		}
		if result.RowsAffected == 0 {
			return errNotFound
		}

		// E-Mail des customer_admin-Users synchronisieren
		if req.Email != nil {
			err := tx.Model(&models.User{}).
				Where("customer_id = ? AND role = ?", id, "customer_admin").
				Updates(map[string]any{"email": *req.Email, "username": *req.Email}).Error
			if err != nil {
				return err
			}
		}
		return nil
	})

	if txErr != nil {
		if errors.Is(txErr, errNotFound) {
			writeError(w, http.StatusNotFound, "Kunde nicht gefunden")
			return
		}
		if strings.Contains(txErr.Error(), "unique") || strings.Contains(txErr.Error(), "UNIQUE") ||
			strings.Contains(txErr.Error(), "duplicate") {
			writeError(w, http.StatusConflict, "E-Mail oder Subdomain wird bereits verwendet")
			return
		}
		writeError(w, http.StatusInternalServerError, "Fehler beim Aktualisieren: "+txErr.Error())
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{"success": true, "message": "Kunde aktualisiert"})
}

// DELETE /api/customers?id=
// Löscht alle abhängigen Daten in korrekter Reihenfolge (FK), dann den Kunden selbst.
func (s *Server) handleDeleteCustomer(w http.ResponseWriter, r *http.Request) {
	idStr := r.URL.Query().Get("id")
	if idStr == "" {
		writeError(w, http.StatusBadRequest, "ID erforderlich")
		return
	}
	id, err := strconv.ParseUint(idStr, 10, 64)
	if err != nil || id == 0 {
		writeError(w, http.StatusBadRequest, "Ungültige ID")
		return
	}

	txErr := s.DB.Transaction(func(tx *gorm.DB) error {
		// Kampagnen-IDs ermitteln, damit campaign-scoped Tabellen gelöscht werden können
		var campaignIDs []uint64
		if err := tx.Model(&models.Campaign{}).
			Where("customer_id = ?", id).
			Pluck("id", &campaignIDs).Error; err != nil {
			return err
		}

		if len(campaignIDs) > 0 {
			// Kampagnen-scoped Tabellen löschen
			for _, table := range []string{"spin_pool", "spins", "leads", "segments", "settings"} {
				if err := tx.Exec("DELETE FROM "+table+" WHERE campaign_id IN ?", campaignIDs).Error; err != nil {
					return err
				}
			}
		}

		// Kampagnen selbst löschen
		if err := tx.Where("customer_id = ?", id).Delete(&models.Campaign{}).Error; err != nil {
			return err
		}

		// User löschen
		if err := tx.Where("customer_id = ?", id).Delete(&models.User{}).Error; err != nil {
			return err
		}

		// Kunden löschen
		result := tx.Delete(&models.Customer{}, id)
		if result.Error != nil {
			return result.Error
		}
		if result.RowsAffected == 0 {
			return errNotFound
		}
		return nil
	})

	if txErr != nil {
		if errors.Is(txErr, errNotFound) {
			writeError(w, http.StatusNotFound, "Kunde nicht gefunden")
			return
		}
		writeError(w, http.StatusInternalServerError, "Fehler beim Löschen: "+txErr.Error())
		return
	}

	writeJSON(w, http.StatusOK, map[string]any{"success": true, "message": "Kunde gelöscht"})
}

// ─── Hilfsfunktionen ──────────────────────────────────────────────────────────

// errNotFound ist ein Sentinel-Fehler für "0 rows affected" in Transaktionen.
var errNotFound = errors.New("not found")

// createDefaultSegments legt die vier Standard-Segmente für eine Kampagne an.
// Port von createDefaultSegments() aus customers.php; in Go auf campaign_id gescoped.
func createDefaultSegments(tx *gorm.DB, campaignID uint64) error {
	type segDef struct {
		name     string
		color    string
		maxCount int
		theme    string
	}
	defaults := []segDef{
		{"Hauptgewinn", "#FFD700", 1, "gold"},
		{"Gewinn", "#4CAF50", 3, "nature"},
		{"Trostpreis", "#2196F3", 5, "water"},
		{"Nochmal drehen", "#FF9800", 3, "sand"},
	}
	for i, d := range defaults {
		seg := models.Segment{
			CampaignID: campaignID,
			Name:       d.name,
			Color:      d.color,
			WinText:    d.name,
			Weight:     100,
			SortOrder:  i,
			MaxCount:   d.maxCount,
			Theme:      d.theme,
			IsActive:   true,
		}
		if err := tx.Create(&seg).Error; err != nil {
			return err
		}
	}
	return nil
}

// createDefaultSettings legt die Standard-Settings für eine Kampagne an.
// Port von createDefaultSettings() aus customers.php; in Go auf campaign_id gescoped.
func createDefaultSettings(tx *gorm.DB, campaignID uint64) error {
	defaults := map[string]string{
		"primary_color":     "#1E3A8A",
		"secondary_color":   "#F6A7C4",
		"accent_color":      "#D4AF37",
		"font_family":       "Montserrat",
		"wheel_title":       "Glücksrad",
		"theme":             "dove",
		"total_spins_limit": "1000",
	}
	for k, v := range defaults {
		setting := models.Setting{
			CampaignID: campaignID,
			Key:        k,
			Value:      v,
		}
		// INSERT OR IGNORE → OnConflict DoNothing
		if err := tx.Clauses(clause.OnConflict{DoNothing: true}).Create(&setting).Error; err != nil {
			return err
		}
	}
	return nil
}

// generatePassword erzeugt ein kryptografisch sicheres Zufalls-Passwort
// aus dem Zeichensatz a-z A-Z 0-9 !@#$%^&* mit der gewünschten Länge.
func generatePassword(length int) (string, error) {
	const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*"
	result := make([]byte, length)
	for i := range result {
		n, err := rand.Int(rand.Reader, big.NewInt(int64(len(charset))))
		if err != nil {
			return "", err
		}
		result[i] = charset[n.Int64()]
	}
	return string(result), nil
}

// isValidEmail prüft eine E-Mail-Adresse auf grundlegende Gültigkeit.
func isValidEmail(email string) bool {
	atIdx := strings.Index(email, "@")
	if atIdx < 1 {
		return false
	}
	domain := email[atIdx+1:]
	return strings.Contains(domain, ".") && len(domain) > 2
}
